streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java

```diff
@@ -253,17 +290,168 @@ public void buildAndOptimizeTopology() {
+        }
+    }
+
+    private void maybeUpdateKeyChangingRepartitionNodeMap() {
+        final Map<StreamsGraphNode, Set<StreamsGraphNode>> mergeNodesToKeyChangers = new HashMap<>();
+        for (final StreamsGraphNode mergeNode : mergeNodes) {
+            mergeNodesToKeyChangers.put(mergeNode, new HashSet<>());
+            final Collection<StreamsGraphNode> keys = keyChangingOperationsToOptimizableRepartitionNodes.keySet();
+            for (final StreamsGraphNode key : keys) {
+                final StreamsGraphNode maybeParentKey = findParentNodeMatching(mergeNode, node -> node.parentNodes().contains(key));
+                if (maybeParentKey != null) {
+                    mergeNodesToKeyChangers.get(mergeNode).add(key);
+                }
+            }
         }
+
+        for (final Map.Entry<StreamsGraphNode, Set<StreamsGraphNode>> entry : mergeNodesToKeyChangers.entrySet()) {
+            final StreamsGraphNode mergeKey = entry.getKey();
+            final Collection<StreamsGraphNode> keyChangingParents = entry.getValue();
+            final Set<OptimizableRepartitionNode> repartitionNodes = new HashSet<>();
+            for (final StreamsGraphNode keyChangingParent : keyChangingParents) {
+                repartitionNodes.addAll(keyChangingOperationsToOptimizableRepartitionNodes.get(keyChangingParent));
+                keyChangingOperationsToOptimizableRepartitionNodes.remove(keyChangingParent);
+            }
+
+            keyChangingOperationsToOptimizableRepartitionNodes.put(mergeKey, repartitionNodes);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private OptimizableRepartitionNode createRepartitionNode(final String name,
+                                                             final Serde keySerde,
+                                                             final Serde valueSerde) {
+
+        final OptimizableRepartitionNode.OptimizableRepartitionNodeBuilder repartitionNodeBuilder = OptimizableRepartitionNode.optimizableRepartitionNodeBuilder();
+        KStreamImpl.createRepartitionedSource(this,
+                                              keySerde,
+                                              valueSerde,
+                                              name + "-optimized",
+                                              name,
+                                              repartitionNodeBuilder);
+
+        return repartitionNodeBuilder.build();
+
+    }
+
+    private StreamsGraphNode getKeyChangingParentNode(final StreamsGraphNode repartitionNode) {
+        final StreamsGraphNode shouldBeKeyChangingNode = findParentNodeMatching(repartitionNode, n -> n.isKeyChangingOperation() || n.isValueChangingOperation());
+
+        final StreamsGraphNode keyChangingNode = findParentNodeMatching(repartitionNode, StreamsGraphNode::isKeyChangingOperation);
+        if (shouldBeKeyChangingNode != null && shouldBeKeyChangingNode.equals(keyChangingNode)) {
```

Comment:
Since the criterion is StreamsGraphNode::isKeyChangingOperation, I don't see why the call on line 407 is needed: if there is no key changing operation, null would be returned anyway.

Reply:
Hmm.. that's a good question. @bbejeck could you take a look?

Reply:
Yep, I'll take a look and push a minor PR if needed.

Reply:
It is needed.  

When we attempt to optimize a key-changing operation, if there are any KStream operations that change the value, we don't do the optimization.   

The call on 407 will retrieve either a value changing or key changing operation from the repartition node.  

If the two retrieved nodes aren't equal on line 410, then we won't perform the optimization for this individual repartition node, as the values could have changed types.

Until we have serde inheritance in place, we can't safely repartition due to serialization/deserialization issues with the new value type.

But I should add some comments to the method to describe what is going on.

Does this make sense?

Reply:
I see what you intended.

Thanks for the response.